src/core/vector/vector.cpp

```diff
@@ -0,0 +1,253 @@
+    add(*this, rhs);
+    return *this;
+}
+
+template<typename ValueType>
+Vector<ValueType>& Vector<ValueType>::operator-=(const Vector<ValueType>& rhs)
+{
+    validateOtherVector(rhs);
+    sub(*this, rhs);
+    return *this;
+}
+
+template<typename ValueType>
+Vector<ValueType> Vector<ValueType>::operator*(const Vector<ValueType>& rhs)
+    requires requires(ValueType a, ValueType b) { a* b; }
+{
+    validateOtherVector(rhs);
+    Vector<ValueType> result(exec_, size_);
+    result = *this;
+    mul(result, rhs);
+    return result;
+}
+
+template<typename ValueType>
+Vector<ValueType> Vector<ValueType>::operator*(const ValueType rhs)
+    requires requires(ValueType a, ValueType b) { a* b; }
+{
+    Vector<ValueType> result(exec_, size_);
+    result = *this;
+    scalarMul(result, rhs);
+    return result;
+}
+
+template<typename ValueType>
+Vector<ValueType>& Vector<ValueType>::operator*=(const Vector<ValueType>& rhs)
+    requires requires(ValueType a, ValueType b) { a *= b; }
+{
+    validateOtherVector(rhs);
+    Vector<ValueType>& result = *this;
+    mul(result, rhs);
+    return result;
+}
+
+template<typename ValueType>
+Vector<ValueType>& Vector<ValueType>::operator*=(const ValueType rhs)
+    requires requires(ValueType a, ValueType b) { a *= b; }
+{
+    Vector<ValueType>& result = *this;
+    scalarMul(result, rhs);
+    return result;
+}
+
+template<typename ValueType>
+void Vector<ValueType>::resize(const localIdx size)
+{
+    void* ptr = nullptr;
+    if (!empty())
+    {
+        std::visit(
+            [this, &ptr, size](const auto& exec)
+            { ptr = exec.realloc(this->data_, static_cast<size_t>(size) * sizeof(ValueType)); },
+            exec_
+        );
+    }
+    else
+    {
+        std::visit(
+            [&ptr, size](const auto& exec)
+            { ptr = exec.alloc(static_cast<size_t>(size) * sizeof(ValueType)); },
+            exec_
+        );
+    }
+    data_ = static_cast<ValueType*>(ptr);
+    size_ = size;
+}
+
+template<typename ValueType>
+void Vector<ValueType>::validateOtherVector(const Vector<ValueType>& rhs) const
+{
+    NF_DEBUG_ASSERT(size() == rhs.size(), "Vectors are not the same size.");
+    NF_DEBUG_ASSERT(exec() == rhs.exec(), "Executors are not the same.");
+}
+
+template<typename ValueType>
+[[nodiscard]] Vector<ValueType> operator+(Vector<ValueType> lhs, const Vector<ValueType>& rhs)
+{
+    lhs += rhs;
+    return lhs;
+}
+
+template<typename ValueType>
+[[nodiscard]] Vector<ValueType> operator-(Vector<ValueType> lhs, const Vector<ValueType>& rhs)
+{
+    lhs -= rhs;
+    return lhs;
+}
+
+// template class instantiation
+template class Vector<uint32_t>;
+template class Vector<uint64_t>;
+template class Vector<int32_t>;
+template class Vector<int64_t>;
+template class Vector<float>;
+template class Vector<double>;
+template class Vector<Vec3>;
+
+// operator instantiation
+#define OPERATOR_INSTANTIATION(Type)                                                               \
```

Comment:
Lets make it like this to help finding it in the future
```suggestion
#define NN_DECLARE_VECTOR_OPERATOR(Type)                                                               \
```